### PR TITLE
MdeModulePkg/Core: CoreValidateHandle optimization

### PR DESCRIPTION
**Pull request by XiaoqiangZhang1**
# Description

<_Include a description of the change and why this change was made._>

<_For each item, place an "x" in between `[` and `]` if true. Example: `[x]` (you can also check items in GitHub UI)_>

<_Create the PR as a Draft PR if it is only created to run CI checks._>

<_Delete lines in \<\> tags before creating the PR._>

- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested

Tested on latest server 8S system

## Integration Instructions

N/A

### CODE REVIEW

**Comment by lgao4** [comment]
Please separate the change for MdePkg as the single commit.

This is good performance enhancement. Do you plan to catch this change in the comming edk2 stable tag 202408?

**Comment by XiaoqiangZhang1** [comment]
> N/A

Thanks Liming,
1. Separated the change for MdePkg:  https://github.com/tianocore/edk2/pull/6063
2. How to address the Spell Check error in MdeModulePkg? add failing words to MdeModulePkg.ci.yaml file?

**Comment by lgao4** [comment]
> > N/A
> 
> Thanks Liming,
> 
> 1. Separated the change for MdePkg:  [MdePkg: CoreValidateHandle Optimization Part1 #6063](https://github.com/tianocore/edk2/pull/6063)
> 2. Yes, if no objection to this patch, please help to merge the MdePkg change first and then I can rebase the MdeModulePkg change to merge.

I mean to add two commits in these Pull Reqest. One is the change in MdePkg, another is in MdeModulePkg. I don't request two PRs.

**Comment by XiaoqiangZhang1** [comment]
> > > N/A
> > 
> > 
> > Thanks Liming,
> > 
> > 1. Separated the change for MdePkg:  [MdePkg: CoreValidateHandle Optimization Part1 #6063](https://github.com/tianocore/edk2/pull/6063)
> > 2. Yes, if no objection to this patch, please help to merge the MdePkg change first and then I can rebase the MdeModulePkg change to merge.
> 
> I mean to add two commits in these Pull Reqest. One is the change in MdePkg, another is in MdeModulePkg. I don't request two PRs.

Thank you, done!

**Comment by lgao4** [comment]
> > > > N/A
> > > 
> > > 
> > > Thanks Liming,
> > > 
> > > 1. Separated the change for MdePkg:  [MdePkg: CoreValidateHandle Optimization Part1 #6063](https://github.com/tianocore/edk2/pull/6063)
> > > 2. Yes, if no objection to this patch, please help to merge the MdePkg change first and then I can rebase the MdeModulePkg change to merge.
> > 
> > 
> > I mean to add two commits in these Pull Reqest. One is the change in MdePkg, another is in MdeModulePkg. I don't request two PRs.
> 
> Thank you, done!

Please remove part1, part2 in the commit message.

**Comment by lgao4** [comment]
@mdkinney， this change requests to catch the stable tag 202408. Have you any comments for it?

**Comment by mdkinney** [comment]
> @mdkinney， this change requests to catch the stable tag 202408. Have you any comments for it?

This is an optimization and not a bug fix.  What is the justification?

**Comment by XiaoqiangZhang1** [comment]
> > > > > N/A
> > > > 
> > > > 
> > > > Thanks Liming,
> > > > 
> > > > 1. Separated the change for MdePkg:  [MdePkg: CoreValidateHandle Optimization Part1 #6063](https://github.com/tianocore/edk2/pull/6063)
> > > > 2. Yes, if no objection to this patch, please help to merge the MdePkg change first and then I can rebase the MdeModulePkg change to merge.
> > > 
> > > 
> > > I mean to add two commits in these Pull Reqest. One is the change in MdePkg, another is in MdeModulePkg. I don't request two PRs.
> > 
> > 
> > Thank you, done!
> 
> Please remove part1, part2 in the commit message.

Done, thank you!

**Comment by lgao4** [comment]
> > @mdkinney， this change requests to catch the stable tag 202408. Have you any comments for it?
> 
> This is an optimization and not a bug fix. What is the justification?

@XiaoqiangZhang1 ， please provide your detail reason. Does this change is critical for your project?

**Comment by XiaoqiangZhang1** [comment]
> > > @mdkinney， this change requests to catch the stable tag 202408. Have you any comments for it?
> > 
> > 
> > This is an optimization and not a bug fix. What is the justification?
> 
> @XiaoqiangZhang1 ， please provide your detail reason. Does this change is critical for your project?

Thanks Liming and Michael , after confirmation, it is okay to not catch the stable tag 202408. The change is mainly for 8S server. You can decide the timeline.

File: MdeModulePkg/Core/Dxe/Hand/Handle.c [hunk 1]
```diff
@@ -59,23 +114,16 @@ CoreValidateHandle (
...
-  for (Link = gHandleList.BackLink; Link != &gHandleList; Link = Link->BackLink) {
-    Handle = CR (Link, IHANDLE, AllHandles, EFI_HANDLE_SIGNATURE);
-    if (Handle == (IHANDLE *)UserHandle) {
-      return EFI_SUCCESS;
-    }
-  }
-
-  return EFI_INVALID_PARAMETER;
+  Entry = OrderedCollectionFind (gOrderedHandleList, UserHandle);
+  return Entry ? EFI_SUCCESS : EFI_INVALID_PARAMETER;
```
Comment (niruiyu) [inline MdeModulePkg/Core/Dxe/Hand/Handle.c:126]: if (Entry == NULL) {
  return EFI_INVALID_PARAMETER;
}
return EFI_SUCCESS;

Comment (XiaoqiangZhang1) [reply MdeModulePkg/Core/Dxe/Hand/Handle.c]: Thanks Ray, changed as following to align with original code logical

  if (Entry != NULL) {
    return EFI_SUCCESS;
  }

  return EFI_INVALID_PARAMETER;

**Review by lgao4 — APPROVED** [review]

**Comment by mdkinney** [comment]
@Mergify rebase

**Pull request merged**